caravan.lua

```diff
@@ -1,31 +1,311 @@
--- Adjusts properties of caravans
---[====[
+-- Adjusts properties of caravans and provides overlay for enhanced trading
+--@ module = true
+
+local gui = require('gui')
+local overlay = require('plugins.overlay')
+local widgets = require('gui.widgets')
+
+trader_selected_state = trader_selected_state or {}
+broker_selected_state = broker_selected_state or {}
+handle_ctrl_click_on_render = handle_ctrl_click_on_render or false
+handle_shift_click_on_render = handle_shift_click_on_render or false
+
+dfhack.onStateChange.caravanOverlay = function(code)
+    if code == SC_WORLD_UNLOADED then
+        trader_selected_state = {}
+        broker_selected_state = {}
+        handle_ctrl_click_on_render = false
+        handle_shift_click_on_render = false
+    end
+end
 
-caravan
-=======
+local GOODFLAG = {
+    UNCONTAINED_UNSELECTED = 0,
+    UNCONTAINED_SELECTED = 1,
+    CONTAINED_UNSELECTED = 2,
+    CONTAINED_SELECTED = 3,
+    CONTAINER_COLLAPSED_UNSELECTED = 4,
+    CONTAINER_COLLAPSED_SELECTED = 5,
+}
 
-Adjusts properties of caravans on the map. See also `force` to create caravans.
+function is_trade_overlay_active()
+    return df.global.game.main_interface.trade.open
+end
 
-This script has multiple subcommands. Commands listed with the argument
-``[IDS]`` can take multiple caravan IDs (see ``caravan list``). If no IDs are
-specified, then the commands apply to all caravans on the map.
+CaravanOverlay = defclass(CaravanOverlay, overlay.OverlayWidget)
```

Comment:
Honestly, this class doesn't seem be adding much value as a subview. I think this class's methods should be integrated into CaravanInfoOverlay and this class can just be removed. If you do want to keep it as a separate class, though, it should inherit from something appropriate, like `gui.View`. It's not an `overlay.OverlayWidget` since it's not directly registered with the overlay framework, and therefore the ATTRS attached to this class will not be read by anything.

Reply:
Just got this one left, should be able to finish it up tonight.